### ISSUE TITLE
Optimize Row Limiting Clause in Oracle Database 12c

### ISSUE THREAD

**Issue by stchr** [doctrine/dbal#2649]
In [12c Release 1 (12.1.0.1)](https://docs.oracle.com/database/121/SQLRF/release_changes.htm#SQLRF56315) Oracle introduced [row_limiting_clause](https://docs.oracle.com/database/121/SQLRF/statements_10002.htm#BABHFGAA), which is faster and easier to understand.

Doctrine\DBAL\Platforms\OraclePlatform::doModifyLimitQuery() should be overwritten with new Doctrine\DBAL\Platforms\Oracle12c1Platform::doModifyLimitQuery()

**Comment by Ocramius** [comment]
> faster and easier to understand.

Unless there is a functional need for a change, I suggest letting `doModifyLimitQuery` untouched even for newer OracleDB versions: that piece of logic is pandora's box, and it's good enough if it works reliably.

**Comment by stchr** [comment]
```php
protected function doModifyLimitQuery($query, $limit, $offset = null)
    {
        if ($limit === null) {
            return $query;
        }
        $limit = (int)$limit;
        $offset = (int)$offset;

        if (preg_match('/^\s*SELECT/i', $query)) {
            if (!preg_match('/\sFROM\s/i', $query)) {
                $query .= ' FROM dual';
            }
            if ($limit > 0) {
                $query .= ' OFFSET ' . $offset . ' ROWS FETCH NEXT ' . $limit . ' ROWS ONLY';
            }
        }

        return $query;
    }
```

vs. 

```php
protected function doModifyLimitQuery($query, $limit, $offset = null)
    {
        $limit = (int) $limit;
        $offset = (int) $offset;

        if (preg_match('/^\s*SELECT/i', $query)) {
            if (!preg_match('/\sFROM\s/i', $query)) {
                $query .= " FROM dual";
            }
            if ($limit > 0) {
                $max = $offset + $limit;
                $column = '*';
                if ($offset > 0) {
                    $min = $offset + 1;
                    $query = 'SELECT * FROM (SELECT a.' . $column . ', rownum AS doctrine_rownum FROM (' .
                            $query .
                            ') a WHERE rownum <= ' . $max . ') WHERE doctrine_rownum >= ' . $min;
                } else {
                    $query = 'SELECT a.' . $column . ' FROM (' . $query . ') a WHERE ROWNUM <= ' . $max;
                }
            }
        }

        return $query;
    }
```

**Comment by Ocramius** [comment]
That still means ending up with two implementations that need to be maintained: can be considered when the supported version of OracleDB is *ALWAYS* `> 12c Release 1 (12.1.0.1)`

**Comment by stchr** [comment]
Ok, thanks.

**Comment by rjd22** [comment]
@Ocramius could we maybe revisit this? I ran into trouble with combining row locking with ROWNUM and using ROWS FETCH might maybe solve this issue.

I will need to do some testing but would supporting more modern Oracle syntax in maybe a separate driver be open for discussion since I'm running into a lot of issues with compatibility with the old syntax.

Example of issue with locking in symfony/messenger

https://github.com/symfony/symfony/pull/45714

**Comment by Ocramius** [comment]
@rjd22 send a patch: I need to be paid looooooooots of money to move even a finger, when something mentions "Oracle" in it.

**Comment by rjd22** [comment]
@Ocramius I hear you. I feel lots of pain working with oracle daily. We want to switch but these are now the cards I've been dealt :(.

I will make a patch. Would making a separate driver be recommended to keep backwards compatibility, like done with MySQL?

**Comment by Ocramius** [comment]
Possibly: beware that I'm no longer the maintainer here, so I endorse creating a clear issue that goes to the point, before throwing more work at it.

**Comment by stchr** [comment]
Would be glad to help if needed 👍

**Comment by rjd22** [comment]
After checking it seems the fix of it is already in the next major version of dbal. So with dbal 4.* This will be solved.

**Comment by stchr** [comment]
Just for reference: This was fixed with #5150
